### PR TITLE
Electron and Displaced tracking

### PR DESCRIPTION
**Pull request by wpmccormack**
PR to DarkQuest e1039 fork which contains:
-Tracking that will work for electrons by removing furthest upstream hit requirements
-Tracking that works for displaced tracks by removing backwards extrapolation in st2+3 to st1 tracklet matching and checks both charge possibilities when tracklet matching
-Flags for turning on both of these.  THEY ARE OFF BY DEFAULT
-Station 3 tracklet information is now preserved in a similar way to full tracks.  This could be useful for particle ID or reconstructing charged particles that are created after the end of station 1 .  Fake rate in realistic simulation or data needs to be studied.

### CODE REVIEW

File: simulation/g4detectors/SQDigitizer.cc [hunk 1]
```diff
@@ -160,7 +160,7 @@ void SQDigitizer::digitizePlane(const std::string& detName)
     const PHG4Hit& g4hit = *(it->second);
 
     int track_id = g4hit.get_trkid();
-    if(track_id < 0) continue;         //only save primary track hits
+    //if(track_id < 0) continue;         //only save primary track hits //WPM
```
Comment (cmantill) [inline simulation/g4detectors/SQDigitizer.cc:163]: Can you comment on why was this needed for your studies? There was actually a recent change that introduces a `digitize_secondaries` flag and does not apply this cut when that flag is on - so I will modify the code to behave like this. But would be good to know in which cases we want to digitize the secondaries.

File: packages/reco/interface/FastTracklet.cxx [hunk 2]
```diff
@@ -734,7 +734,17 @@ double Tracklet::getMomentum() const
 
 int Tracklet::getCharge() const
 {
-	return x0*KMAGSTR > tx ? 1 : -1;
+  if(!_chargeSet){
```
[the remark this answers is not in the document]
Comment (wpmccormack) [reply packages/reco/interface/FastTracklet.cxx:737]: The new formula shouldn't affect my results

File: packages/reco/ktracker/KalmanFastTracking.cxx [hunk 3]
```diff
@@ -812,7 +821,22 @@ void KalmanFastTracking::buildGlobalTracks()
                     resolveLeftRight(tracklet_global, 150.);
                     resolveSingleLeftRight(tracklet_global);
                 }
-
+		if(TRACK_DISPLACED){
+		  double firstChiSq = tracklet_global.calcChisq();
+		  Tracklet tracklet_global2 = (*tracklet23) * (*tracklet1);
+		  tracklet_global2.setCharge(-1*tracklet_global2.getCharge());
```
[the remark this answers is not in the document]
Comment (cmantill) [reply packages/reco/ktracker/KalmanFastTracking.cxx]: Ok thanks, sorry I didn't ask the question properly, my question really is: why is the inversion of the charge of `tracklet_global2 ` needed?

File: simulation/g4detectors/SQDigitizer.cc (hunk 1 above)
Comment (wpmccormack) [reply simulation/g4detectors/SQDigitizer.cc:163]: So for particle guns, if there's any showering of the initial particle, any hits associated with secondaries are not included in the tracker hit record.  One thing that we're trying to look at is particle ID, so if there's e.g. a pair production from a photon or brem from an electron, then tracks won't get built for the secondary particles unless the secondaries are digitized.  Hits for the secondaries do occur in the ECal even if only primary track hits are saved, but not in the tracker

Comment (cmantill) [reply simulation/g4detectors/SQDigitizer.cc:163]: Ok thanks, this is useful to know! For now, we will keep this `digitize_secondaries` function in the merged PR, and maybe we can set it to always `True` in our `Fun4Sim.C` macro.

File: packages/reco/ktracker/KalmanFastTracking.cxx [hunk 4]
```diff
@@ -78,6 +78,8 @@ namespace
 
     //if displaced, skip fit to the target/vertex
     static bool NOT_DISPLACED;
+    static bool TRACK_ELECTRONS;
```
[the remark this answers is not in the document]
Comment (wpmccormack) [reply packages/reco/ktracker/KalmanFastTracking.cxx]: I added comments in framework/phool/recoConsts.cc, which is where the flags are defined, and I added small comments here pointing to that file for further reference

**Review by cmantill — APPROVED** [review]

**Pull request merged**